src/NServiceBus.Core/Routing/UnicastRoutingTable.cs

```diff
@@ -46,24 +57,21 @@ public void AddStatic(Type messageType, EndpointInstanceName destination)
         /// <param name="destinationAddress">Destination endpoint instance address.</param>
         public void AddStatic(Type messageType, string destinationAddress)
         {
-            rules.Add((t, c) => StaticRule(t, messageType, new UnicastRoute(destinationAddress)));
+            staticRules.Add((t, c) => StaticRule(t, messageType, new UnicastRoute(destinationAddress)));
         }
 
         /// <summary>
         /// Adds a rule for generating unicast routes.
         /// </summary>
-        /// <param name="dynamicRule">The rule.</param>
-        public void AddDynamic(Func<Type, ContextBag, IEnumerable<IUnicastRoute>> dynamicRule)
+       // /// <param name="dynamicRule">The rule.</param>
+        public void AddDynamic(Func<Type, ContextBag, Task<List<IUnicastRoute>>> dynamicRule)
```

Comment:
Not 100% sure but I think we need a `Func<Task<List<IUnicastRoute>>` or a Lazy. Because when you accept a Task I think it will be scheduled already when you add it. But in your design you want to resolve them every time asynchronously when you acquired the routes.